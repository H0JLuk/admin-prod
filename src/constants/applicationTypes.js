const applicationTypes = [
    {type: 'IOS'},
    {type: 'ANDROID'},
    {type: 'OTHER'}
];

export default applicationTypes;