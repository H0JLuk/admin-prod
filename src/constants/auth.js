export const LOGIN_TYPE = {
    PASSWORD: 'PASSWORD',
    // SBOL_PRO: 'SBOL_PRO',
    SBER_USER_ID: 'SBER_USER_ID',
};