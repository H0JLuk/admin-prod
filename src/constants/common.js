export const showCount = { formatter: ({ count, maxLength }) => `Осталось символов ${maxLength - count}` };

/* '^(https?:\\/\\/)?'+ // protocol
    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
    '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
    '(\\:\\d+)?(\\/[-a-z\\d%_.~+{}]*)*'+ // port and path
    '(\\?[;&a-z\\d%_.~+=-{}]*)?'+ // query string
    '(\\#[-a-z\\d_]*)?$','i'); // fragment locator */
export const URL_REGEXP = /^(https?:\/\/)?((([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}|((\d{1,3}\.){3}\d{1,3}))(:\d+)?(\/[-A-Za-z\d%_.~+#]*)*(\?[;&a-z\d%_.~+=-{}]*)?(#[-a-z\d_]*)?$/;
