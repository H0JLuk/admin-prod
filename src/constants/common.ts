import { TextAreaProps } from 'antd/lib/input';

export const showCount: TextAreaProps['showCount'] = {
    formatter: ({ count, maxLength }) => `Осталось символов ${maxLength! - count}`,
};

export const DEFAULT_SLEEP_TIME = 500; // in milliseconds

export enum DIRECTION {
    ASC = 'ASC',
    DESC = 'DESC',
}
