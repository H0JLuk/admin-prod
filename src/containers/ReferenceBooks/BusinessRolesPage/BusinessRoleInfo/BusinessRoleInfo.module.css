.container {
    width: 100%;
    padding: 0px 60px 0px 30px;
    position: relative;
}

.wrapper {
    padding: 32px;
    background-color: #f8f8f8;
}

.block {
    max-width: 800px;
}

.title {
    font-weight: 600;
    font-size: 20px;
    line-height: 28px;
    letter-spacing: -.38px;
    margin-top: 24px;
    margin-bottom: 13px;
}

.btnGroup {
    margin: 24px 0;
}

.btnGroup > button {
    margin-right: 15px;
}

.fieldValue {
    font-weight: 600;
}

.error {
    margin-top: 5px;
    color: #dd283b;
}
