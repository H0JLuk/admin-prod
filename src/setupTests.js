import '@testing-library/jest-dom';
import { configure as enzymeConfigure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';
import './jestMocks';

enzymeConfigure({ adapter: new Adapter() });

const globalTimeout = global.setTimeout;

export const sleep = async (timeout = 0) => {
    await act(async () => {
        await new Promise((resolve) => globalTimeout(resolve, timeout));
    });
};
