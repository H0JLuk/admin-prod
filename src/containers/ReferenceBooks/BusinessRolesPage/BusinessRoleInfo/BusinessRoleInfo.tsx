import React, { useState, useEffect, useRef } from 'react';
import { generatePath, RouteComponentProps } from 'react-router-dom';
import { Col, Row, Button } from 'antd';

import Loading from '@components/Loading';
import { getBusinessRoleById, deleteBusinessRole } from '@apiServices/businessRoleService';

import { BusinessRoleDto } from '@types';
import { customNotifications } from '@utils/notifications';
import { confirmModal } from '@utils/utils';
import { BUSINESS_ROLE_PAGES } from '@constants/route';
import { getFormattedDate } from '@utils/helper';
import { BUTTON_TEXT } from '@constants/common';

import styles from './BusinessRoleInfo.module.css';

const BUSINESS_ROLE_INFO_FIELDS: { label: string; name: keyof BusinessRoleDto; }[] = [
    {
        label: 'Название',
        name: 'name',
    },
    {
        label: 'Описание',
        name: 'description',
    },
    {
        label: 'Начало действия',
        name: 'startDate',
    },
];

function getNotifyDeletedMessage(name: string) {
    return <span>Бизнес-роль <b>{name}</b> успешно удалена</span>;
}

interface BusinessRoleInfoProps extends RouteComponentProps<{ businessRoleId: string; }> {
    matchPath: string;
}

const BusinessRoleInfo: React.FC<BusinessRoleInfoProps> = ({
    matchPath,
    match,
    history,
}) => {
    const { businessRoleId } = match.params;
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState('');
    const businessRole = useRef({} as BusinessRoleDto);

    const goToBusinessRolesList = (redirect?: boolean) => history[redirect ? 'replace' : 'push'](matchPath);

    useEffect(() => {
        if (!businessRoleId) {
            goToBusinessRolesList(true);
            return;
        }

        (async () => {
            try {
                const businessRoleData = await getBusinessRoleById(businessRoleId);

                if (!businessRoleData) {
                    goToBusinessRolesList(true);
                    return;
                }

                businessRole.current = {
                    ...businessRoleData,
                    startDate: getFormattedDate(businessRoleData.startDate),
                    endDate:  getFormattedDate(businessRoleData.endDate),
                };
            } catch (err) {
                setError(err.message);
            }
            setLoading(false);
        })();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    if (loading) {
        return (
            <div className={styles.container}>
                <Loading />
            </div>
        );
    }

    const onEdit = () => {
        history.push(generatePath(`${ matchPath }${ BUSINESS_ROLE_PAGES.EDIT_BUSINESS_ROLE }`, { businessRoleId }), { businessRole: businessRole.current });
    };

    const deleteRole = async () => {
        try {
            await deleteBusinessRole(businessRoleId);
            customNotifications.success({
                message: getNotifyDeletedMessage(businessRole.current.name),
            });
            goToBusinessRolesList(true);
        } catch (e) {
            setError(e.message);
        }
    };

    const onDeleteClick = () => {
        confirmModal({
            onOk: deleteRole,
            title: <span>Вы действительно хотите удалить бизнес-роль <b>{businessRole.current.name}</b>?</span>,
        });
    };

    return (
        <div className={styles.container}>
            <div className={styles.title}>
                {businessRole.current.name}
            </div>
            <div className={styles.wrapper}>
                <Row gutter={[24, 30]} className={styles.block}>
                    {BUSINESS_ROLE_INFO_FIELDS.map((field, index) => (
                        <Col
                            key={index}
                            span={12}
                        >
                            <div>{field.label}</div>
                            <div className={styles.fieldValue}>
                                {businessRole.current[field.name] || <i>{`${field.label} отсутствует`}</i>}
                            </div>
                        </Col>
                    ))}
                </Row>
                {error && <span className={styles.error}>{error}</span>}
            </div>
            <div className={styles.btnGroup}>
                <Button
                    type="primary"
                    onClick={onEdit}
                >
                    {BUTTON_TEXT.EDIT}
                </Button>
                <Button
                    type="primary"
                    danger
                    onClick={onDeleteClick}
                >
                    {BUTTON_TEXT.DELETE}
                </Button>
            </div>
        </div>
    );
};

export default BusinessRoleInfo;
