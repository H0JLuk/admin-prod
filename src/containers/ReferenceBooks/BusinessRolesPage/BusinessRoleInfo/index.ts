import BusinessRoleInfo from './BusinessRoleInfo';

export default BusinessRoleInfo;
