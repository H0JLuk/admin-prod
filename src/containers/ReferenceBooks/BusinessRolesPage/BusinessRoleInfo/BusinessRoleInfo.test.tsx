import React from 'react';
import { fireEvent, render, act, screen, waitFor } from '@testing-library/react';
import BusinessRoleInfo from './BusinessRoleInfo';
import {
    deleteBusinessRole,
    getBusinessRoleById,
} from '@apiServices/businessRoleService';
import { sleep } from '../../../../setupTests';
import { BUTTON_TEXT } from '@constants/common';
import { testBusinessRole } from '../../../../../__tests__/constants';

jest.mock('@apiServices/businessRoleService', () => ({
    deleteBusinessRole: jest.fn(),
    getBusinessRoleById: jest.fn(),
}));

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();

const match = (id: string | null = '1') => ({
    params: {
        businessRoleId: id,
    },
});

const BusinessRolesListTestProps: any = {
    matchPath: 'test',
    history: {
        push: mockHistoryPush,
        replace: mockHistoryReplace,
    },
};

const errorMessage = 'error';

describe('BusinessRoleList test', () => {
    const Component = (matchParams = match()) => render(<BusinessRoleInfo {...BusinessRolesListTestProps} match={matchParams} />);

    beforeEach(() => {
        (deleteBusinessRole as jest.Mock).mockResolvedValue({ message: 'Successful', status: 'Ok' });
        (getBusinessRoleById as jest.Mock).mockResolvedValue(testBusinessRole);
    });

    it('should match snapshot', async () => {
        const RenderComponent = Component();
        await sleep();
        expect(RenderComponent.container).toMatchSnapshot();
    });

    it('should delete business role', async () => {
        await waitFor(() => {
            Component();
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.DELETE));
        });

        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText(/Да/));
        });

        expect(deleteBusinessRole).toBeCalledTimes(1);
    });

    it('should delete business role with error', async () => {
        (deleteBusinessRole as jest.Mock).mockRejectedValue({ message: errorMessage });

        await waitFor(() => {
            Component();
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.DELETE));
        });

        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText(/Да/));
        });

        expect(deleteBusinessRole).toBeCalledTimes(1);
        expect(screen.getByText(errorMessage)).toBeTruthy();
    });

    it('should redirect to edit business role', async () => {
        await waitFor(() => {
            Component();
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.EDIT));
        });

        await sleep();

        expect(mockHistoryPush).toBeCalledTimes(1);
    });

    it('should call get role with mock api function', async () => {
        await waitFor(() => {
            Component();
        });

        expect(getBusinessRoleById).toBeCalledTimes(1);
    });

    it('should not find business role id and call error', async () => {
        (getBusinessRoleById as jest.Mock).mockRejectedValue({ message: errorMessage });

        await waitFor(() => {
            Component();
        });

        expect(screen.getByText(errorMessage)).toBeTruthy();
    });

    it('should not find business role id', async () => {
        await waitFor(() => {
            Component(match(null));
        });

        expect(mockHistoryReplace).toBeCalledTimes(1);
    });

    it('should return error for business role id', async () => {
        (getBusinessRoleById as jest.Mock).mockResolvedValue('');

        await waitFor(() => {
            Component();
        });

        expect(mockHistoryReplace).toBeCalledTimes(1);
    });

});
