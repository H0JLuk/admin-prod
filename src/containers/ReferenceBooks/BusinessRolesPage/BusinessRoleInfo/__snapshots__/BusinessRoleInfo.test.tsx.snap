// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessRoleList test should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      Мобильный менеджер по продажам
    </div>
    <div
      class="wrapper"
    >
      <div
        class="ant-row block"
        style="margin: -15px -12px -15px -12px;"
      >
        <div
          class="ant-col ant-col-12"
          style="padding: 15px 12px 15px 12px;"
        >
          <div>
            Название
          </div>
          <div
            class="fieldValue"
          >
            Мобильный менеджер по продажам
          </div>
        </div>
        <div
          class="ant-col ant-col-12"
          style="padding: 15px 12px 15px 12px;"
        >
          <div>
            Описание
          </div>
          <div
            class="fieldValue"
          >
            description
          </div>
        </div>
        <div
          class="ant-col ant-col-12"
          style="padding: 15px 12px 15px 12px;"
        >
          <div>
            Начало действия
          </div>
          <div
            class="fieldValue"
          >
            <i>
              Начало действия отсутствует
            </i>
          </div>
        </div>
      </div>
      
    </div>
    <div
      class="btnGroup"
    >
      <button
        class="ant-btn ant-btn-primary"
        type="button"
      >
        <span>
          Редактировать
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-dangerous"
        type="button"
      >
        <span>
          Удалить
        </span>
      </button>
    </div>
  </div>
</div>
`;
